jormungandr/src/metrics/backends/prometheus_exporter.rs

```diff
@@ -0,0 +1,248 @@
+    }
+
+    fn set_tip_block(
+        &self,
+        block: &chain_impl_mockchain::block::Block,
+        block_ref: &crate::blockchain::Ref,
+    ) {
+        let mut block_tx_count = 0;
+        let mut block_input_sum = Value::zero();
+        let mut block_fee_sum = Value::zero();
+
+        block
+            .contents
+            .iter()
+            .try_for_each::<_, Result<(), ValueError>>(|fragment| {
+                fn totals<T>(t: &Transaction<T>) -> Result<(Value, Value), ValueError> {
+                    Ok((t.total_input()?, t.total_output()?))
+                }
+
+                let (total_input, total_output) = match &fragment {
+                    Fragment::Transaction(tx) => totals(tx),
+                    Fragment::OwnerStakeDelegation(tx) => totals(tx),
+                    Fragment::StakeDelegation(tx) => totals(tx),
+                    Fragment::PoolRegistration(tx) => totals(tx),
+                    Fragment::PoolRetirement(tx) => totals(tx),
+                    Fragment::PoolUpdate(tx) => totals(tx),
+                    Fragment::VotePlan(tx) => totals(tx),
+                    Fragment::VoteCast(tx) => totals(tx),
+                    Fragment::VoteTally(tx) => totals(tx),
+                    Fragment::EncryptedVoteTally(tx) => totals(tx),
+                    Fragment::Initial(_)
+                    | Fragment::OldUtxoDeclaration(_)
+                    | Fragment::UpdateProposal(_)
+                    | Fragment::UpdateVote(_) => return Ok(()),
+                }?;
+                block_tx_count += 1;
+                block_input_sum = (block_input_sum + total_input)?;
+                let fee = (total_input - total_output).unwrap_or_else(|_| Value::zero());
+                block_fee_sum = (block_fee_sum + fee)?;
+                Ok(())
+            })
+            .expect("should be good");
+
+        self.block_tx_count.set(block_tx_count);
+        self.block_input_sum.set(block_input_sum.0);
+        self.block_fee_sum.set(block_fee_sum.0);
+        self.block_content_size
+            .set(block.header.block_content_size().into());
+        self.block_epoch.set(block.header.block_date().epoch.into());
+        self.block_slot
+            .set(block.header.block_date().slot_id.into());
+        let chain_length: u32 = block.header.chain_length().try_into().unwrap();
+        self.block_chain_length.set(chain_length as u64);
+        self.block_time.set(
+            block_ref
+                .time()
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap()
+                .as_secs(),
+        );
+
+        let block_hash = block.header.hash();
+        let block_hash_bytes = block_hash.as_bytes();
+        for i in 0..4 {
+            let mut value_bytes = [0u8; 8];
+            value_bytes.copy_from_slice(&block_hash_bytes[(8 * i)..(8 * (i + 1))]);
+            let value = u64::from_le_bytes(value_bytes);
+            self.block_hash[i].set(value);
```

Comment:
Just wondering, splitting a block hash in 4 atomics u64 prevents us from blocking the thread, but could potentially return invalid values for a brief moment. Maybe we could put a (acq / rel) fence ? Not sure if the library allows such flexibility though

Reply:
That's a good idea, but yeah, there is no such functionality and getting invalid values is possible. I am not sure how to deal with that without introducing a lock. As you mentioned, this library is not very flexible.

Reply:
Another idea could be to store the hash inside the `Prometheus` struct with some other atomic thing like arcswap and then update the registry only inside `http_response` before gathering metrics *if* the block hash has been updated.

Reply:
Yeah, that's what I thought.